google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -289,32 +372,48 @@ private PooledSession(Session delegate) {
       markUsed();
     }
 
+    @VisibleForTesting
+    void setAllowDelegateRecreation(boolean allow) {
+      allowDelegateRecreation = allow;
+    }
+
+    private void maybeRecreateUnderlyingSession(SessionNotFoundException e) {
+      if (beforeFirstServerCall && allowDelegateRecreation) {
+        delegate = spanner.createSession(db);
+      } else {
+        throw e;
+      }
+    }
+
     private void markBusy() {
       this.state = SessionState.BUSY;
       this.leakedException = new LeakedSessionException();
+      this.beforeFirstServerCall = true;
     }
 
     private void markClosing() {
       this.state = SessionState.CLOSING;
     }
 
-    @Override
-    public Timestamp write(Iterable<Mutation> mutations) throws SpannerException {
+    /**
+     * Executes a method that does a database call and retries the method if a {@link
+     * SessionNotFoundException} is thrown.
+     */
+    private <T> T runWithSessionRetry(Supplier<T> method) {
       try {
-        markUsed();
-        return delegate.write(mutations);
-      } catch (SpannerException e) {
-        throw lastException = e;
-      } finally {
-        close();
+        return internalRunWithSessionRetry(method);
+      } catch (SessionNotFoundException e) {
+        maybeRecreateUnderlyingSession(e);
+        lastException = null;
+        return internalRunWithSessionRetry(method);
```

Comment:
there is a chance we got sessionNotFound here as well.. 

have u considered putting this in a loop ? 
Pros: customers never see a SessionNotFound
Cons: high latency, but our timeouts should still take effect.